gemma/run.cc

```diff
@@ -188,20 +188,20 @@ void ReplGemma(const ThreadingArgs& threading, const InferenceArgs& inference,
                                     .use_spinning = threading.spin};
     inference.CopyTo(runtime_config);
     std::vector<int> prompt;
-    size_t prompt_size = 0;
     size_t prefix_end = 0;
     if (have_image) {
       prompt = WrapAndTokenize(gemma.Tokenizer(), gemma.ChatTemplate(),
                                config.wrapping, abs_pos, prompt_string,
                                image_tokens.Rows());
       runtime_config.image_tokens = &image_tokens;
       prompt_size = prompt.size();
-      // The end of the prefix for prefix-LM style attention in Paligemma.
-      // See Figure 2 of https://arxiv.org/abs/2407.07726.
-      prefix_end = prompt_size;
-
-      // REMOVED: Don't change prefill_tbatch_size for image handling
-      // runtime_config.prefill_tbatch_size = prompt_size;
+      if (config.wrapping == PromptWrapping::PALIGEMMA) {
+        // The end of the prefix for prefix-LM style attention in Paligemma.
+        // See Figure 2 of https://arxiv.org/abs/2407.07726.
+        prefix_end = prompt_size;
```

Comment:
The prefix_end change makes sense, thanks.
But the prefill_tbatch_size change can result in huge batch sizes and should not be necessary. Results ought to be the same regardless of batch size. Can we revert that?

What I can imagine is that we should still call activations.SetBatchSize. This sets the (active) number of rows in each tensor, which is initially the batch size passed to the Activations ctor.

Reply:
The intention was to split up the matmuls because at some point, larger batch sizes are slower, and for the matmul itself, it doesn't matter whether we chunk it or do all at once. I agree that Prefix-LM wants to attend to all image tokens. What we are missing for the chunking to work is the online softmax also used for flash attention, which is on the roadmap.
So you are right, it seems for the moment our attention has to have all K vectors for all tokens. Let's add a comment mentioning this and the plan to support online softmax, after which we can lift this requirement.